### ISSUE TITLE
`ResolverFn` should potentially include array of promises

### ISSUE THREAD

**Issue by bnaoki** [dotansimha/graphql-code-generator#1965]
**Describe the bug**
 
This is regarding the `typescript-resolvers` plugin.

At least with Apollo Server, a resolver is able to return an array of promises in addition to a single promise that resolves to an array. Right now, the generated type for `ResolverFn` supports only the latter option. Should it be updated to handle the array of promises as well?

https://www.apollographql.com/docs/apollo-server/essentials/data/#resolver-results
> A promise - resolvers often do asynchronous actions like fetching from a database or backend API, so they can return promises. This can be combined with arrays, so a resolver can return a promise that resolves to an array, or an array of promises, and both are handled correctly.

**Comment by dobesv** [comment]
Not to mention you can probably return a promise of an array of promises.

**Comment by dotansimha** [comment]
@kamilkisiela can you please add it? I think we need to change the signature of `ResolverFn`.

**Comment by kamilkisiela** [comment]
Work in progress: https://github.com/dotansimha/graphql-code-generator/pull/2023

**Comment by kamilkisiela** [comment]
@dobesv @bnaoki could you guys try `1.2.2-alpha-1a7ee826.35+1a7ee826`?

**Comment by dotansimha** [comment]
Fixed in 1.3.0 🎉

**Comment by kjvalencik** [comment]
Hi! Thanks for building this feature. I encountered an edge case with the types generated that this fix does not cover.

If you have a resolver for an array of items and each of those items is nullable, not all valid return values are possible in the type.

```ts
// GraphQL: items(): [Item]!

// Works
type ItemsResolver = () => (Item | null)[];

// Works
type ItemsResolver = () => Promise<(Item | null)[]>;

// Works
type ItemsResolver = () => (Promise<Item> | null)[];

// Does not work
type ItemsResolver = () => (Promise<Item | null>)[];
```

The last use case is especially useful when providing a resolver that accepts multiple IDs, some of which may fail (for example, the `nodes` resolver in Relay).

As a workaround, we can cast (either explicitly or with a non-null assertion) to an inaccurate (but, valid) type.

```ts
const promises: Promise<Item | null)[];

promises as Promise<Item>[];
```

Using `Promise.all` is not a work-around because it is semantically different. If any single promise fails, the entire array will be `null `instead of the individual element.

I looked into the cause of this issue, but the fix is not obvious. The query resolver is referencing the resolver types in order to support the mapping feature (`Array<Maybe<ResolversTypes['Item']>>`). However, the `ResolverTypes` is performing the wrapping for sync/async resolution. A correct fix may require introducing a new intermediate type.

Let me know if I can provide any additional details or if you have a suggested implementation. Thanks!
